text/0000-with-bounds-in-pi-types.md

```diff
@@ -0,0 +1,685 @@
+     const upper_b: usize,
+     const lower_b: usize> ops::Add<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     // We have to satisfy the constraints set out in the struct definition.
+     with lower_a <= upper_a,
+           lower_b <= upper_b,
+           // Check for overflow by some `const fn`.
+           is_overflow_safe(upper_a, upper_b) {
+    // These parameters are constant expression.
+    type Output = BoundedInt<lower_a + lower_b, upper_a + upper_b>;
+
+    fn add(self, rhs: BoundedInt<lower_b, upper_b>) -> Self::Output {
+        BoundedInt {
+            n: self.n + rhs.n,
+        }
+    }
+}
+
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> From<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     with lower_a <= upper_a,
+           lower_b <= upper_b,
+           // We will only extend the bound, never shrink it without runtime
+           // checks, thus we add this clause:
+           lower_b <= lower_a && upper_b >= upper_a {
+    fn from(from: BoundedInt<lower_b, upper_b>) -> Self {
+        BoundedInt {
+            n: from.n,
+        }
+    }
+}
+```
+
+### Homogeneous varargs
+
+We can use arbitrarily length arrays to simulate homogeneous varargs:
+
+```rust
+fn my_func<const n: usize>(args: [u32; n]) { /* whatever */ }
+
+my_func([1, 2, 3]);
+my_func([1, 2, 3, 4]);
+```
+
+### Array generics
+
+Currently libcore only implements various traits up to arrays of length 32.
+This allows for implementing them for arrays of arbitrary length:
+
+```rust
+impl<const n: usize, T: Clone> Clone for [T; n] {
+    fn clone(&self) -> [T; n] {
+        // Clone it...
+    }
+}
+```
+
+### Statically checked indexing
+
+One can perform simple, interval based, statically checked indexing:
+
+```rust
+use std::ops;
+
+impl<const n: usize, T: Clone> ops::Index<BoundedInt<0, n - 1>> for [T; n] {
+    type Output = T;
+
+    fn index(&self, ind: BoundedInt<0, n - 1>) -> &T {
+        unsafe {
+            // This is safe due to the bound on `ind`.
+            self.unchecked_index(*ind)
+        }
+    }
+}
+```
+
+### Fancy number stuff
+
+```rust
+struct Num<const n: usize>;
+
+trait Divides<const m: usize> {}
+
+impl<const a: usize, const b: usize> Divides<b> for Num<a> with b % a == 0 {}
+```
+
+### Playground
+
+[This repo](https://github.com/ticki/rfc-1657-playground) is a playground for
+usecases of such a feature. Refer to that for more examples.
+
+## `with` clauses
+
+Often, it is wanted to have some statically checked clause satisfied by the
+constant parameters (e.g., for the sake of compile-time bound checking). To
+archive this, in a reasonable manner, we use constexprs, returning a boolean.
+
+We allow such constexprs in `with` clauses of functions. Whenever the
+function is invoked given constant parameters `<a, b...>`, the compiler
+evaluates this expression, and if it returns `false`, an aborting error is
+invoked.
```

Comment:
Does "an aborting error is invoked" mean we give up on compilation entirely, or go back to checking for other impls a la SFINAE? This wording sounds like it'd be a compilation error, but I would've expected that we want to allow "impl Foo<n> for 0<n<=2" to coexist alongside "impl Foo<n> for n>2".